File: Projects/ShopifyLarkTracker/ShopifyLarkTracker/Services/ShipmentProcessor.cs
using ShopifyLarkTracker.Models;
using System.Text.Json;

namespace ShopifyLarkTracker.Services;

public class ShipmentProcessor
{
    private readonly ExceptionEngine _rules = new();
    private readonly LarkBaseService _lark;

    public ShipmentProcessor(
        LarkBaseService lark)
    {
        _lark = lark;
    }

    public async Task ProcessAsync(
        List<ShopifyOrder> orders,
        JsonDocument larkDoc)
    {
        var root =
            larkDoc.RootElement;

        if (root.GetProperty("code")
                .GetInt32() != 0)
        {
            Console.WriteLine(
                "Failed loading Lark records");

            return;
        }

        var items =
            root.GetProperty("data")
                .GetProperty("items");

        var recordMap =
            BuildRecordMap(items);

        Console.WriteLine(
            $"Loaded {recordMap.Count} existing Lark records");

        var semaphore =
            new SemaphoreSlim(50);

        var tasks =
            new List<Task>();

        foreach (var order in orders)
        {
            var shipment = new Shipment
            {
                ShopifyId =
                    order.ShopifyId,

                OrderNo =
                    order.OrderNumber,

                Channel =
                    NormalizeChannel(
                        order.SourceName),

                CustomerName =
                    order.ShippingAddress?.Name
                    ??
                    $"{order.Customer?.FirstName} {order.Customer?.LastName}"
                        .Trim(),

                Status =
                    order.Status,

                OrderCreatedDate =
                    order.CreatedAt,

                LastUpdated =
                    DateTime.Now
            };

            if (order.Fulfillments?.Any() == true)
            {
                shipment.TrackingNo =
                    order.Fulfillments[0]
                        .TrackingNumber;

                shipment.FulfillmentDate =
                    order.Fulfillments[0]
                        .CreatedAt;
            }

            _rules.Evaluate(shipment);

            if (string.IsNullOrWhiteSpace(
                shipment.OrderNo))
            {
                continue;
            }

            recordMap.TryGetValue(
                shipment.OrderNo,
                out var recordId);

            tasks.Add(
                ProcessOrderAsync(
                    shipment,
                    recordId,
                    semaphore));
        }

        await Task.WhenAll(tasks);

        Console.WriteLine(
            $"Finished processing {tasks.Count} orders");
    }

    private async Task ProcessOrderAsync(
        Shipment shipment,
        string? recordId,
        SemaphoreSlim semaphore)
    {
        await semaphore.WaitAsync();

        try
        {
            var fields =
                BuildFields(shipment);

            if (string.IsNullOrWhiteSpace(
                    recordId))
            {
                Console.WriteLine(
                    $"CREATE -> {shipment.OrderNo}");

                await _lark.CreateRecordAsync(
                    fields);
            }
            else
            {
                Console.WriteLine(
                    $"UPDATE -> {shipment.OrderNo}");

                await _lark.UpdateRecordAsync(
                    recordId,
                    fields);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine(
                $"ERROR [{shipment.OrderNo}] : {ex.Message}");
        }
        finally
        {
            semaphore.Release();
        }
    }

    private Dictionary<string, string>
        BuildRecordMap(
            JsonElement items)
    {
        var map =
            new Dictionary<string, string>(
                StringComparer.OrdinalIgnoreCase);

        foreach (var item in
            items.EnumerateArray())
        {
            if (!item.TryGetProperty(
                    "fields",
                    out var fields))
            {
                continue;
            }

            if (!fields.TryGetProperty(
                    "订单号",
                    out var orderField))
            {
                continue;
            }

            var orderNo =
                orderField.ToString()?.Trim();

            if (string.IsNullOrWhiteSpace(
                    orderNo))
            {
                continue;
            }

            var recordId =
                item.GetProperty("record_id")
                    .GetString();

            if (!string.IsNullOrWhiteSpace(
                    recordId))
            {
                map[orderNo] =
                    recordId;
            }
        }

        return map;
    }

    private string NormalizeChannel(
        string? source)
    {
        source =
            source?.Trim()
                  .ToLower()
            ?? "";

        return source switch
        {
            "web" => "Online Store",
            "pos" => "POS",
            _ => source
        };
    }

    private Dictionary<string, object>
        BuildFields(
            Shipment shipment)
    {
        return new Dictionary<string, object>
        {
            ["订单号"] =
                shipment.OrderNo,

            ["客户名称"] =
                shipment.CustomerName ?? "",

            ["快递单号"] =
                shipment.TrackingNo ?? "",

            ["运单号"] =
                shipment.ShopifyId.ToString(),

            ["物流公司"] =
                GetCourier(
                    shipment.TrackingNo),

            ["签收状态"] =
                string.IsNullOrWhiteSpace(
                    shipment.Status)
                    ? "Pending"
                    : shipment.Status,

            ["多选"] =
                string.IsNullOrWhiteSpace(
                    shipment.Channel)
                ? Array.Empty<string>()
                : new[]
                {
                    shipment.Channel
                },

            ["是否异常"] =
                shipment.IsException
                    ? "Yes"
                    : "No",

            ["发货日期"] =
                shipment.FulfillmentDate?
                    .ToString("yyyy/MM/dd")
                ??
                shipment.OrderCreatedDate?
                    .ToString("yyyy/MM/dd")
                ?? "",

            ["AWB创建时间"] =
                shipment.FulfillmentDate?
                    .ToString(
                        "yyyy/MM/dd HH:mm:ss")
                ??
                shipment.OrderCreatedDate?
                    .ToString(
                        "yyyy/MM/dd HH:mm:ss")
                ?? "",

            ["更新时间"] =
                DateTime.UtcNow.ToString(
                    "yyyy-MM-ddTHH:mm:ssZ"),

            ["备注"] =
                shipment.Remarks ?? ""
        };
    }

    private string GetCourier(
        string? trackingNo)
    {
        if (string.IsNullOrWhiteSpace(
                trackingNo))
        {
            return "To Ship";
        }

        trackingNo =
            trackingNo.Trim()
                .ToUpper();

        if (trackingNo.StartsWith("SPX"))
            return "SPX";

        if (trackingNo.StartsWith("SP"))
            return "SPX";

        if (trackingNo.StartsWith("JT"))
            return "J&T";

        return "Unknown";
    }
}

File: Projects/ShopifyLarkTracker/logivis-worker/src/services/ShipmentProcessor.ts
import { ExceptionEngine } from "./ExceptionEngine";
import { LarkBaseService, larkBaseService } from "./lark-base.service";
import { larkAuthService } from "./lark-auth.service";
import type { ShopifyOrder } from "../models/shopify";
import type { ShipmentSync } from "../models/ShipmentSync";
import type { ListRecordsResponse } from "../models/lark-record";

export interface BatchUpdateShopifyRecord {
	record_id: string;
	fields: Record<string, any>;
}

export class ShipmentProcessor {

	private readonly rules = new ExceptionEngine();

	constructor(
		private env: Env,
		private lark: LarkBaseService
	) {
	}

	public async processAsync(
		orders: ShopifyOrder[],
		larkDoc: ListRecordsResponse
	): Promise<void> {

		const runId = crypto.randomUUID();

		console.log(`Run ${runId} started`);
		console.log(`Processing ${orders.length} Shopify orders...`);

		const token =
			await larkAuthService.getTenantAccessToken(this.env);

		const items = larkDoc.data.items;

		const recordMap =
			this.buildRecordMap(items);

		console.log(
			`Loaded ${recordMap.size} existing Lark records`
		);

		const seen = new Set<string>();

		for (const order of orders) {

			const key =
				this.normalizeOrderNo(order.orderNumber);

			if (seen.has(key)) {

				console.warn(
					`Duplicate Shopify order: ${key}`
				);

			}

			seen.add(key);

		}

		// Batch collections
		const updates: BatchUpdateShopifyRecord[] = [];
		const creates: { fields: Record<string, any> }[] = [];

		for (const order of orders) {

			const shipment: ShipmentSync = {

				shopifyId:
					order.shopifyId,

				orderNo:
					order.orderNumber,

				channel:
					this.normalizeChannel(
						order.sourceName
					),

				customerName:
					order.shippingAddress?.name ??
					`${order.customer?.firstName ?? ""} ${order.customer?.lastName ?? ""}`.trim(),

				status:
					order.status,

				isException:
					false,

				orderCreatedDate:
					order.createdAt,

				lastUpdated:
					new Date()
			};

			if (order.fulfillments?.length > 0) {

				shipment.trackingNo =
					order.fulfillments[0].trackingNumber;

				shipment.fulfillmentDate =
					order.fulfillments[0].createdAt;

			}

			this.rules.evaluate(shipment);

			if (!this.normalizeOrderNo(shipment.orderNo)) {
				continue;
			}

			const recordId =
				recordMap.get(
					this.normalizeOrderNo(
						shipment.orderNo
					)
				);

			if (recordId) {

				updates.push({
					record_id: recordId,
					fields: this.buildUpdateFields(shipment)
				});

			}
			else {

				creates.push({
					fields: this.buildCreateFields(shipment)
				});

			}

		}

		const start = Date.now();

		await this.lark.batchUpdateRecordsShopify(
			this.env,
			token,
			updates
		);

		await this.lark.batchCreateRecords(
			this.env,
			token,
			creates
		);

		console.log(
			`Elapsed = ${Date.now() - start} ms`
		);

		console.log(
			`Updated ${updates.length} records`
		);

		console.log(
			`Created ${creates.length} records`
		);

		console.log(
			`Finished processing ${orders.length} orders`
		);

	}

	
	private buildRecordMap(items: any[]): Map<string, string> {

		const map = new Map<string, string>();

		for (const item of items) {

			const fields = item.fields;

			if (!fields) {
				console.warn(`Record ${item.record_id} has no fields.`);
				continue;
			}

			const rawOrderNo = fields["订单号"];
			const orderNo = this.normalizeOrderNo(rawOrderNo);

			if (!orderNo) {
				console.warn(
					`Record ${item.record_id} has blank OrderNo.`,
					rawOrderNo
				);
				continue;
			}

			if (map.has(orderNo)) {

				console.warn(
					`Duplicate OrderNo: ${orderNo}`,
					{
						recordId: item.record_id,
						rawOrderNo
					}
				);

				continue;
			}

			map.set(orderNo, item.record_id);

		}

		console.log(
			`Map Size = ${map.size}, Lark Records = ${items.length}`
		);

		return map;

	}


	private normalizeOrderNo(value: unknown): string {

		return value
			?.toString()
			.replace(/^#/, "")
			.trim() ?? "";

	}

	private normalizeChannel(
		source?: string
	): string {

		source =
			source?.trim().toLowerCase() ?? "";

		switch (source) {

			case "web":
				return "Online Store";

			case "pos":
				return "POS";

			default:
				return source;
		}

	}

	private async processOrderAsync(
		token: string,
		shipment: ShipmentSync,
		recordId?: string
	): Promise<void> {

		try {

			if (!recordId) {

				console.log(
					`CREATE -> ${shipment.orderNo}`
				);

				const fields = this.buildCreateFields(shipment);

				// TODO: Port CreateRecord
				await this.lark.createRecord(
					this.env,
					token,
					fields
				);

			}
			else {

				console.log(
					`UPDATE -> ${shipment.orderNo}`
				);

				const fields = this.buildUpdateFields(shipment);

				await this.lark.updateRecordShopify(
					this.env,
					token,
					recordId,
					fields
				);

			}

		}
		catch (ex) {

			console.error(
				`ERROR [${shipment.orderNo}]`,
				ex
			);

		}

	}

	private buildCreateFields(
		shipment: ShipmentSync
	): Record<string, any> {

		return {
			"订单号": shipment.orderNo,
			"客户名称": shipment.customerName ?? "",
			"快递单号": shipment.trackingNo ?? "",
			"运单号": shipment.shopifyId.toString(),
			"物流公司": this.getCourier(shipment.trackingNo),
			"签收状态": shipment.status?.trim() || "Pending",
			"多选": shipment.channel ? [shipment.channel] : [],
			"是否异常": shipment.isException ? "Yes" : "No",
			"发货日期": shipment.fulfillmentDate ?? shipment.orderCreatedDate,
			"AWB创建时间": shipment.fulfillmentDate ?? shipment.orderCreatedDate,
			"更新时间": shipment.lastUpdated.toISOString(),
			"备注": shipment.remarks ?? ""
		};

	}

	private buildUpdateFields(
		shipment: ShipmentSync
	): Record<string, any> {

		return {
			"订单号": shipment.orderNo,
			"客户名称": shipment.customerName ?? "",
			"发货日期":
				(shipment.fulfillmentDate ?? shipment.orderCreatedDate)?.toISOString(),
			"AWB创建时间":
				(shipment.fulfillmentDate ?? shipment.orderCreatedDate)?.toISOString(),
			"运单号":
				shipment.shopifyId.toString()
		};

	}

	private async runWithConcurrency(
		tasks: Array<() => Promise<void>>,
		concurrency: number
	): Promise<void> {

		console.log(
			`[Scheduler] Total Tasks = ${tasks.length}, Concurrency = ${concurrency}`
		);

		const executing: Promise<void>[] = [];

		let started = 0;

		for (const task of tasks) {

			started++;

			console.log(
				`[Scheduler] Starting Task ${started}/${tasks.length} | Active = ${executing.length}`
			);

			const promise = task().finally(() => {

				const index = executing.indexOf(promise);

				if (index >= 0) {
					executing.splice(index, 1);
				}

				console.log(
					`[Scheduler] Task Finished | Active = ${executing.length}`
				);

			});

			executing.push(promise);

			if (executing.length >= concurrency) {

				console.log(
					`[Scheduler] Waiting... Active = ${executing.length}`
				);

				await Promise.race(executing);

				console.log(
					`[Scheduler] Resumed | Active = ${executing.length}`
				);

			}

		}

		console.log(
			`[Scheduler] Waiting for remaining ${executing.length} tasks...`
		);

		await Promise.all(executing);

		console.log(
			`[Scheduler] All tasks completed.`
		);

	}

	private getCourier(
		trackingNo?: string
	): string {

		if (!trackingNo?.trim()) {
			return "To Ship";
		}

		trackingNo = trackingNo.trim().toUpperCase();

		if (
			trackingNo.startsWith("SPX") ||
			trackingNo.startsWith("SP")
		) {
			return "SPX";
		}

		if (
			trackingNo.startsWith("JT")
		) {
			return "J&T";
		}

		return "Unknown";

	}
}
